cas-server-support-osf/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java

```diff
@@ -177,20 +197,30 @@ protected Event doExecute(final RequestContext context) throws Exception {
             restoreRequestAttribute(request, session, LOCALE);
             restoreRequestAttribute(request, session, METHOD);
 
-            // 5. Attempt to authenticate if the credential is not null
             if (credentials != null) {
-                final TicketGrantingTicket tgt =
-                        this.centralAuthenticationService.createTicketGrantingTicket(new ClientCredential(credentials));
+                // 5. Attempt to authenticate if the credential is not null.
+                final TicketGrantingTicket tgt = this.centralAuthenticationService
+                        .createTicketGrantingTicket(new ClientCredential(credentials));
                 WebUtils.putTicketGrantingTicketInScopes(context, tgt);
                 return success();
+            } else {
+                // Otherwise, abort the authentication: prepare the login context with clients info and then go to the
+                // original start point of the CAS login web flow.
+                prepareForLoginPage(context);
+                return error();
             }
-        }
 
-        // If `client_name` is not in the request params - no auth, or if `credentials == null` - aborted auth, simply
-        // prepare the login context with clients info and then go to the original start point of the CAS login web
-        // flow: refer to "cas-server-webapp/src/main/webapp/WEB-INF/webflow/login/login-webflow.xml" for detail.
-        prepareForLoginPage(context);
-        return error();
+        } catch (final Exception e) {
+            // If any exception occurs during the client authentication process, return the specialized event which
+            // will allow the CAS authentication exception handler to handle the exception gracefully.
+            final Map<String, Class<? extends Exception>> failures = new LinkedHashMap<>();
+            failures.put(e.getClass().getSimpleName(), e.getClass());
+            return getEventFactorySupport().event(
+                    this,
+                    AUTHENTICATION_FAILURE,
+                    new LocalAttributeMap<>("error", new AuthenticationException(failures))
+            );
+        }
```

Comment:
Turn all exceptions into an event `AUTHENTICATION_FAILURE` for the login web flow to redirect to the primary CAS authentication handler. The exception detail is stored in the event attributes.